<a href="https://colab.research.google.com/github/IsraelAlcAlv/proyecto_final_fake_news/blob/master/Proyecto_M%C3%B3dulo_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Fake News y su Impacto en la Sociedad** 

Las noticias falsas o inventadas (fake news) han marcado la historia de diferentes maneras. Algunos las pueden llegar a identificar fácilmente, pero a muchos otros les costará verificar su veracidad y las tomarán como ciertas. Esto representa un gran problema para la sociedad, ya que polariza y aísla a los diferentes grupos sociales dependiendo de sus gustos.


Medios actuales, investigadores, instituciones y demás entidades  han desarrollado o propuesto diversas alternativas para la solución de las dichosas Fake News, van desde consejos para no ser víctima de alguna información falsa o inverosímil, hasta modelos matemáticos y computacionales más complejos para detectar y/o predecir la ocurrencia de una, entre estos últimos,  se han desarrollado diversos modelos de clasificación  usando Support Vector Machine (SVM) y Naive Bayes Classifier (NBC). También se utilizan la Regresión Logística (LR) y los modelos basados en Árboles de Decisiones.


## **Preguntas que queremos responder**





* ¿Cómo diferenciamos entre noticias falsas y noticias reales?
* ¿Podemos predecir la aparición de estas noticias?
* ¿Se presentan en alguna estacionalidad?
* ¿Qué porcentaje de noticias falsas abundan hoy en día en los medios de comunicación?
* ¿Qué lugar/es tiene mayor abundancia de estas noticias?
* ¿Qué temas son los más comunes entre las noticias falsas?
* ¿Qué impacto tiene en la sociedad?
* ¿Qué porcentaje de la población toma las noticias falsas como reales?
* ¿Qué medios de comunicación tienen mayor número de noticias falsas?

# **Análisis preliminar de los datos**

Las noticias recolectadas en este dataset cubren el transcurso de fechas entre el 25/10/16 al 25/11/16

En este dataset existen artículos con información verídica tanto como información engañosa mejor conocida como “fake news”

Las noticias en este dataset provienen de una variedad de países como lo son USA, Canadá, Colombia, Rusia, Australia, Francia entre otros, sin embargo la gran mayoría son originarias del primero mencionado.

Dado el periodo de tiempo tan específico de esta colección y el origen de estas noticias  podemos suponer que gran parte de las mismas giran en torno a un evento que popularizara el término “fake news” siendo este las elecciones de USA en 2016 y su resultado final con Donald Trump como el siguiente presidente electo para esta nación.


## **Importar datos**



La base de datos se extrajo de Kaggle.

Las noticias recolectadas en este dataset cubren el transcurso de fechas entre el 25/10/16 al 25/11/16.

En este dataset existen artículos con información verídica tanto como información engañosa mejor conocida como “fake news”

https://www.kaggle.com/fanbyprinciple/fake-news-dataset

El dataset se encuentra en la siguiente carpeta:

https://drive.google.com/drive/folders/1GN1Bt3Kr007xGoaIGesmvxqKQBainFEU?usp=sharing

Agregando un acceso directo a nuestra unidad de drive podrá acceder directamente a los datos.

df = pd.read_csv("/content/drive/MyDrive/FakeNews/fake_news_dataset.csv")

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

## **Descarga del DataSet**

Con la librería **pandas** y la función **read_csv** es posible descargar la base de datos para trabajar.

Importamos la librería **pandas**, una librería de Python especializada en el manejo y análisis de estructuras de datos.

Para leer un archivo .csv en pandas, usamos **read_csv**

In [ ]:
# Descargando dataset
import pandas as pd
df = pd.read_csv("/content/drive/MyDrive/4_BEDU/Data Science/Python/fake_news_dataset.csv")

Con **head** vemos los nombres de las columnas y un número determinado de filas:

In [ ]:
# Primer vistazo a la información
df.head()

uuid  ord_in_thread  ... shares  type
0  6a175f46bcd24d39b3e962ad0f29936721db70db              0  ...      0  bias
1  2bdc29d12605ef9cf3f09f9875040a7113be5d5b              0  ...      0  bias
2  c70e149fdd53de5e61c29281100b9de0ed268bc3              0  ...      0  bias
3  7cf7c15731ac2a116dd7f629bd57ea468ed70284              0  ...      0  bias
4  0206b54719c7e241ffe0ad4315b808290dbe6c0f              0  ...      0  bias

[5 rows x 20 columns]

Para conocer más información sobre las columnas se usa **info**.

Información general de todas las columnas, aquí podremos ver si los tipos correspondientes a cada columna son válidos para continuar o si requieren modificación, además del nombre de cada columna y el total de registros.

In [ ]:
 df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12999 entries, 0 to 12998
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   uuid                12999 non-null  object 
 1   ord_in_thread       12999 non-null  int64  
 2   author              10575 non-null  object 
 3   published           12999 non-null  object 
 4   title               12319 non-null  object 
 5   text                12953 non-null  object 
 6   language            12999 non-null  object 
 7   crawled             12999 non-null  object 
 8   site_url            12999 non-null  object 
 9   country             12823 non-null  object 
 10  domain_rank         8776 non-null   float64
 11  thread_title        12987 non-null  object 
 12  spam_score          12999 non-null  float64
 13  main_img_url        9356 non-null   object 
 14  replies_count       12999 non-null  int64  
 15  participants_count  12999 non-null  int64  
 16  like

## **Descripción general de las columnas**

Las noticias son extraídas de Twitter y se clasifican según el hilo donde se publicaron, título, autor, lenguaje, país y la interacción que tuvo ese tweet (likes, comentarios, shares) y una columna muy importante, type. Esta nos brinda información valiosa sobre la naturaleza de la noticia, específicamente podemos extraer de aquí las Fake News. Ya que no existe una descripción detallada de las columnas nosotros concluimos que:


*   uuid se usa como identificador único para cada noticia.
*   ord_in_thread como el orden en el hilo en el cual aparece la noticia.
*   site_url como la url del portal de origen de la noticia.
*   thread_title como el título del hilo.
*   crawled, domain_rank, spam_score, main_img_url al no tener una descripción *   intuitiva decidimos no utilizarlas.
*   Las demás columnas en sí mismas se explican y se sobreentiende su uso.



Datos como replies_count, participants_count, likes, comments y shares nos pueden dar un gran indicativo de que noticias fueron las que causaron más impacto en esta fechas y nos dan la capacidad para inferir de cierta manera sobre las noticias.

## **Limpieza de datos**

De la base de datos seleccionamos las columnas que nos dan información relevante para el caso de estudio

In [ ]:
# Selección de columnas significantes
columns = ["author","published","title","text","language","country","replies_count","participants_count","likes","comments","shares","type"]
clean_df = df[columns]
print(clean_df.head())
clean_df.shape

                 author                      published  ... shares  type
0     Barracuda Brigade  2016-10-26T21:41:00.000+03:00  ...      0  bias
1  reasoning with facts  2016-10-29T08:47:11.259+03:00  ...      0  bias
2     Barracuda Brigade  2016-10-31T01:41:49.479+02:00  ...      0  bias
3                Fed Up  2016-11-01T05:22:00.000+02:00  ...      0  bias
4                Fed Up  2016-11-01T21:56:00.000+02:00  ...      0  bias

[5 rows x 12 columns]


(12999, 12)

### **NaNs**

Se usa la función vectorizada  isna para contar los NaNs (regresa True cuando encuentra un NaN y False cuando el valor es válido).

Con sum podemos contar cuántos NaNs existen, que suma 1 por cada True y 0 por cada False:

In [ ]:
#hacemos un conteo preliminar para conocer el numero de na existentes por columna
clean_df.isna().sum()

author                2424
published                0
title                  680
text                    46
language                 0
country                176
replies_count            0
participants_count       0
likes                    0
comments                 0
shares                   0
type                     0
dtype: int64

### **Eliminar NaNs**

Observamos que la cantidad de NaNs no es tan grande, por la que podemos eliminarnos sin problema alguno. Con dropna se hace la limpieza de NaNs y nos quedan datos completamente útiles para nuestro análisis.

In [ ]:
#eliminamos los na y comprobamos que no quede ninguno
no_na_df = clean_df.dropna()
print(no_na_df.isna().sum())
print(no_na_df.shape)

author                0
published             0
title                 0
text                  0
language              0
country               0
replies_count         0
participants_count    0
likes                 0
comments              0
shares                0
type                  0
dtype: int64
(9809, 12)


### **Reindexando**

Una vez limpio nuestro dataset debemos de reindexar nuestros registros para conservar un buen orden de nuestros datos. Esto lo logramos con reset_index utilizando el parámetro de drop igualado a True para quedarnos con los nuevos indices y deshacernos de los indices anteriores.


In [ ]:
no_na_df = no_na_df.reset_index(drop=True)

Usando **info** y **tail** para revisar la nueva base de datos sin NaNs

In [ ]:
no_na_df.info()
no_na_df.tail()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9809 entries, 0 to 9808
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   author              9809 non-null   object
 1   published           9809 non-null   object
 2   title               9809 non-null   object
 3   text                9809 non-null   object
 4   language            9809 non-null   object
 5   country             9809 non-null   object
 6   replies_count       9809 non-null   int64 
 7   participants_count  9809 non-null   int64 
 8   likes               9809 non-null   int64 
 9   comments            9809 non-null   int64 
 10  shares              9809 non-null   int64 
 11  type                9809 non-null   object
dtypes: int64(5), object(7)
memory usage: 919.7+ KB


author                      published  ... shares type
9804       Tyler Durden  2016-10-26T23:25:49.124+03:00  ...     14   bs
9805       Tyler Durden  2016-10-26T23:53:43.161+03:00  ...     72   bs
9806  George Washington  2016-10-26T23:53:49.879+03:00  ...    333   bs
9807       Tyler Durden  2016-10-27T00:20:00.111+03:00  ...    197   bs
9808       Tyler Durden  2016-10-27T00:35:58.174+03:00  ...    714   bs

[5 rows x 12 columns]

### **Cambio de tipo de datos**

Ya limpio nuestro dataset y reindexado podemos enfocarnos en el tipo de dato. Observamos que la única columna que necesita un cambio de tipo es la columna de published, ya que esta contiene la fecha en la que la noticia fue publicada. El método pandas to_datetime ayuda a convertir un dato tipo string Date time en un objeto Date time para poder utilizarlo de forma correcta.

In [ ]:
pd.to_datetime(no_na_df["published"])

0              2016-10-26 21:41:00+03:00
1       2016-10-29 08:47:11.259000+03:00
2       2016-10-31 01:41:49.479000+02:00
3              2016-11-01 05:22:00+02:00
4              2016-11-01 21:56:00+02:00
                      ...               
9804    2016-10-26 23:25:49.124000+03:00
9805    2016-10-26 23:53:43.161000+03:00
9806    2016-10-26 23:53:49.879000+03:00
9807    2016-10-27 00:20:00.111000+03:00
9808    2016-10-27 00:35:58.174000+03:00
Name: published, Length: 9809, dtype: object

Una vez terminado todo este proceso de limpieza tenemos un dataset completamente funcional y listo para usarse.

# **Filtrando la información, obteniendo subdataframes y ordenando**

Para empezar a constestar algunas de las preguntas iniciales debemos realizar filtros.







## ***-¿Cómo diferenciamos entre noticias falsas y noticias reales?***

Afortunadamente nuestro dataset contiene una columna dedicada a que tipo de noticia es la que estamos observando, asi que lo ideal será investigar que valores tiene esta por medio de ***unique***

In [ ]:
no_na_df['type'].unique()  #['bias', 'conspiracy', 'fake', 'bs', 'hate', 'junksci', 'state']

el resultado es un array donde podemos encontrar el valor "fake".
Esta es una gran ayuda para clasificar nuestra información entre que noticias son reales y cuales no, tomando como reales todos los demas valores del array.

Incluso, podemos obtener un subdataframe que contenga solo noticias fake y contar sus registros.

In [ ]:
fake_df = no_na_df[(no_na_df.type.str.contains("fake"))] #nuevo subdf
print(fake_df)

fake_df.shape #(14, 12) <- 14 filas, 12 columnas

De forma inesperada, el numero de noticias clasificadas como "fake" son solo 14 de 9809 registros.

Si bien este dataset nos brinda una forma muy sencilla de identificar que es considerado como fake news y que no, no siempre es asi y se requiere de análisis profundos y exhaustivos, en conjunto de otras tecnologias como la que se hablará a continuación. 

## ***-¿Podemos predecir la aparición de estas noticias?***

En esta primera etapa del proyecto, no.

Sin embargo, en un futuro, basandonos en la información obtenida a partir de este dataset y con la ayuda de AI, podriamos analizar el texto, la procedencia, las palabras que usan, contextos políticos, fechas, entre otros. Toda esta informacion serviría para alimentar la AI y poder dar una respuesta a esta pregunta.

## ***-¿Qué porcentaje de noticias falsas abundan hoy en día en los medios de comunicación?***

De nuevo, en esta etapa es imposible responder esta pregunta sin ayuda de AI, dicho lo anterior, podemos obtener el porcentaje para este dataframe y el periodo de tiempo que abarca.

In [ ]:
prob_fake = (len(fake_df)/len(no_na_df)*100) #Con len contamos la cantidad de registros en cada df, dividimos fakes entre total y el resultado se multiplica x 100
prob_fake #0.14272606789682946

Dando como resultado un muy pequeño 0.14% de fake news dentro de este conjunto de información.

## ***¿Qué lugar/es tiene mayor abundancia de estas noticias?***

Para obtener esta informacion aplicaremos una funcion de agregación (count) a nuestro fake_df, agrupando por país (groupby). 

In [ ]:
fake_df.groupby("country")["title"].count() 

# country
#CO     1
#US    13

De las 14 fake news tenemos 13 pertenecientes a USA.

Esto, en conjunto con datos como el rango de fechas especificos de este dataset nos empieza a dar pequeños indicios de cual puede ser el tema principal, al menos de este subconjunto especifico de fake news.

## ***¿Qué temas son los más comunes entre las noticias falsas?***

Dado que en este caso específico el subdataframe contiene muy pocos registros, basta con leer los titulos de forma manual, y encontramos un patron en casi todos los titulos, las elecciones de USA-2016

In [ ]:
fake_df['title'].unique() 

Podemos hacer un conteo de las noticias que incluyen en su título algunos terminos relacionados a las elecciones.

Esto usando la propiedad contains, especificando un case insensitive y regex para que busque los terminos en cualquier posicion de los strings correspondientes.

In [ ]:
elections_fake_df = fake_df[fake_df['title'].str.contains('trump|campaing|clinton|white house|president|election', regex=True, na=False, case=False)]
len(elections_fake_df) #9

De las 14 fake news donde 13 se originaron en USA, 9 son en relación a la campaña electoral 2016...mmm... parece que tenemos una tema principal claramente definido, al menos en este subdataset.

## ***¿Qué más se puede hacer con esta información?, ¿Qué tal el impacto?***

Debido a las fechas tan especificas de este dataset y análisis previo, se llegó a una suposición del tema principal y por tanto, con mayor impacto de estas noticias, siendo este el fin de la campaña electoral de Estados Unidos-2016 y la victoria de Trump como presidente electo, podemos comprobar esto con filtros.

Filtrando esta información podemos obtener subdatasets importantes para continuar con el análisis a futuro como los son:

> -Noticias con 500 shares o más (Cuando empiezan a ser relevantes para un mayor público)

> -Noticias que involucran a Donald Trump

> -Noticias que involucran a Donald Trump y tienen más de 500 shares

## **Para encontrar las noticias con más de 500 shares.**

Del DataFrame no_na_df se seleccionan la columna *shares* y sólo se toman las filas con más de 500 y se crea el DataFrame **impact_df**.

Después con **sort** se ordenan por mayor cantidad de shares y respuestas.


In [ ]:
#Noticias con mas de 500 shares
impact_df = no_na_df[(no_na_df.shares > 500)]

impact_df.sort_values(by=['shares', 'replies_count'], ascending= False) #ordenando por mayor cantidad de shares y respuestas

## **Encontrando las noticias que hablen de "Trump"**

Se seleccionan los datos de tipo string (texto) que coincidan con *Trump*

Posteriormente se ordenan con **sort** por mayor cantidad de shares y comentarios

Aqui lo que realmente queremos saber es el total de noticias por lo que usamos la propiedad shape

In [ ]:
#Noticias que contengan "Trump"
trump_df = no_na_df[(no_na_df.text.str.contains("Trump"))]

trump_df.sort_values(by=['shares', 'replies_count'], ascending= False)

trump_df.shape

Con los dos filtros pasados ahora se realiza un filtrado de **noticias que involucran a Trump y tienen más de 500 shares.**







In [ ]:
#Noticias que involucran a Trump y tienen más de 500 shares
trump_imp_df = no_na_df[(no_na_df.text.str.contains("Trump")) & (no_na_df.shares > 500)]

trump_imp_df.sort_values(by=['shares', 'replies_count'], ascending= False)

# **Api The Guardian**

Se realizaron peticiones a una api **The Guardian**, un sitio  de noticias y medios, la api almacena todos los artículos, imágenes, audio y videos que datan de 1999. Todo accesible con una sola clave de plataforma abierta.

Para obtener la documentación https://open-platform.theguardian.com/access/

Primero se necesita el endpoint y diccionario, después se pasan al **método GET de requests** y se realiza la petición a ese endpoint, seguido envía los parámetros, por ejemplo, la información extra que el API necesita para validar la petición:

In [ ]:
import requests
payload = {'api_key' : '4bdcba55-8588-4ec7-b5b9-4ed50fc42aab'}

endpoint = "https://content.guardianapis.com/search?q=Hillary%20AND%20Trump&api-key=4bdcba55-8588-4ec7-b5b9-4ed50fc42aab"
r = requests.get(endpoint, params=payload)


Posteriormente, nuestro objeto de respuesta da el siguiente resultado:

In [ ]:
r.status_code

200

El valor  de **200** indica que todo salió bien con la petición.

Como es una respuesta larga se necesita diseccionarla:

In [ ]:
r.json()


{'response': {'currentPage': 1,
  'orderBy': 'relevance',
  'pageSize': 10,
  'pages': 839,
  'results': [{'apiUrl': 'https://content.guardianapis.com/us-news/2021/mar/07/lucky-review-how-biden-beat-trump-and-doubters-like-obama-and-hillary',
    'id': 'us-news/2021/mar/07/lucky-review-how-biden-beat-trump-and-doubters-like-obama-and-hillary',
    'isHosted': False,
    'pillarId': 'pillar/news',
    'pillarName': 'News',
    'sectionId': 'us-news',
    'sectionName': 'US news',
    'type': 'article',
    'webPublicationDate': '2021-03-07T06:00:55Z',
    'webTitle': 'Lucky review: how Biden beat Trump – and doubters like Obama and Hillary',
    'webUrl': 'https://www.theguardian.com/us-news/2021/mar/07/lucky-review-how-biden-beat-trump-and-doubters-like-obama-and-hillary'},
   {'apiUrl': 'https://content.guardianapis.com/us-news/2021/jun/08/trump-biden-democrats-hillary-clinton-michelle-obama-2020-election-book',
    'id': 'us-news/2021/jun/08/trump-biden-democrats-hillary-clinton-mich

In [ ]:
json['response']['results']

[{'apiUrl': 'https://content.guardianapis.com/books/2021/feb/23/hillary-clinton-to-publish-thriller-set-in-aftermath-of-us-political-turmoil',
  'id': 'books/2021/feb/23/hillary-clinton-to-publish-thriller-set-in-aftermath-of-us-political-turmoil',
  'isHosted': False,
  'pillarId': 'pillar/arts',
  'pillarName': 'Arts',
  'sectionId': 'books',
  'sectionName': 'Books',
  'type': 'article',
  'webPublicationDate': '2021-02-23T13:12:12Z',
  'webTitle': 'Hillary Clinton to publish thriller set in aftermath of US political turmoil',
  'webUrl': 'https://www.theguardian.com/books/2021/feb/23/hillary-clinton-to-publish-thriller-set-in-aftermath-of-us-political-turmoil'},
 {'apiUrl': 'https://content.guardianapis.com/politics/2021/jun/16/keir-starmer-picks-stonger-together-slogan-used-by-hillary-clinton-to-rebrand',
  'id': 'politics/2021/jun/16/keir-starmer-picks-stonger-together-slogan-used-by-hillary-clinton-to-rebrand',
  'isHosted': False,
  'pillarId': 'pillar/news',
  'pillarName': 'N

## Así obtenemos las 10 noticias más relevantes de Hillary y Trump

Convirtiendo a un DataFrame:

In [ ]:
import pandas as pd 
normalized = json['response']['results']
df = pd.DataFrame.from_dict(normalized)
df

id  ... pillarName
0  books/2021/feb/23/hillary-clinton-to-publish-t...  ...       Arts
1  politics/2021/jun/16/keir-starmer-picks-stonge...  ...       News
2  us-news/live/2021/may/06/liz-cheney-republican...  ...       News
3  us-news/2021/mar/07/lucky-review-how-biden-bea...  ...       News
4      books/2020/nov/28/best-politics-books-of-2020  ...       Arts
5  stage/2020/oct/26/the-narcissist-christopher-s...  ...       Arts
6  us-news/2021/jun/08/trump-biden-democrats-hill...  ...       News
7  us-news/2021/may/16/joe-biden-feels-political-...  ...       News
8  commentisfree/2021/feb/16/two-party-electoral-...  ...    Opinion
9  commentisfree/2020/nov/08/american-politics-tr...  ...    Opinion

[10 rows x 11 columns]